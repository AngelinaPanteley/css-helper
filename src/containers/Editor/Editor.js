import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import styles from './Editor.scss';
import { calcStyles } from '../../shared/utility';
import * as actions from '../../store/actions/index';
import Tabs from '../../components/UI/Tabs/Tabs';
import Generator from '../../components/Generator/Generator';
import Preview from '../../components/Preview/Preview';
import Code from '../../components/Code/Code';
import Examples from '../../components/Examples/Examples';
import Auxiliary from '../../hoc/Auxiliary/Auxiliary';
import SaveModal from '../../components/SaveModal/SaveModal';

class Editor extends PureComponent {
  static propTypes = {
    settings: PropTypes.shape({
      controls: PropTypes.object.isRequired,
    }).isRequired,
    isAuth: PropTypes.bool.isRequired,
    userId: PropTypes.string,
    token: PropTypes.string,
    editingItemId: PropTypes.string,
    editingItemTitle: PropTypes.string,
    save: PropTypes.func.isRequired,
    saveEditedItem: PropTypes.func.isRequired,
    openHint: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props);

    const controls = props.settings.controls;
    const initialControlValues = {};
    let controlValues = JSON.parse(localStorage.getItem(this.props.name));

    for (let name in controls) {
      const control = controls[name];
      initialControlValues[name] = control.initialValue;
    };

    if (!controlValues) {
      controlValues = { ...initialControlValues };
    }

    const styleValues = calcStyles(props.settings.style, controlValues, controls);
    const initialStyleValues = calcStyles(props.settings.style, initialControlValues, controls);

    this.state = {
      controlValues,
      initialControlValues,
      styleValues,
      initialStyleValues,
      isExamplesOpen: true,
      isModalOpen: false,
    }
  }

  componentWillUnmount() {
    localStorage.setItem(this.props.name, JSON.stringify(this.state.controlValues));
  }

  handleChange = (control, value) => {
    this.setState({
      controlValues: {
        ...this.state.controlValues,
        [control]: value,
      }
    });

    this.changeStyles();
  }

  changeStyles = () => {
    const settings = this.props.settings;

    this.setState((prevState) => {
      return {
        styleValues: calcStyles(settings.style, prevState.controlValues, settings.controls),
      }
    });
  }

  toggleExamples = (e) => {
    e.preventDefault();
    this.setState({
      isExamplesOpen: !this.state.isExamplesOpen,
    })
  }

  selectExample = (exampleName) => {
    const example = this.props.settings.examples[exampleName];

    this.setState({
      controlValues: {
        ...this.state.initialControlValues,
        ...example,
      }
    });
    this.changeStyles();
  }

  toggleSaveModal = (e) => {
    e.preventDefault();

    this.setState({
      isModalOpen: !this.state.isModalOpen,
    })
  }

  onSave = (title) => {
    const { userId, token } = this.props;
    this.props.save(title, this.props.name, this.state.controlValues, userId, token);
  }

  saveEdited = (e) => {
    e.preventDefault();
    const { userId, token } = this.props;
    this.props.saveEditedItem(
      this.props.editingItemId,
      this.props.editingItemTitle,
      this.props.name,
      this.state.controlValues,
      userId,
      token
    );
  }

  openCopyHint = () => {
    this.props.openHint('Copied.');
  }

  clearControlValues = () => {
    this.setState({
      styleValues: { ...this.state.initialStyleValues },
      controlValues: { ...this.state.initialControlValues },
    })
  }

  render() {
    const settings = this.props.settings;
    return (
      <Auxiliary>
        <div className={styles.Editor}>
          <div className={styles.PreviewContainer}>
            <Examples isOpen={this.state.isExamplesOpen}
              previewClass={settings.className}
              template={settings.template}
              examples={settings.examples}
              initialControlValues={this.state.initialControlValues}
              styles={settings.style}
              controls={settings.controls}
              selectExample={this.selectExample} />
            <a href="" onClick={this.toggleExamples} className={styles.ToggleLink}>
              {
                this.state.isExamplesOpen
                  ?
                  'Hide Examples'
                  :
                  'Show Examples'
              }
            </a>
            <div className={styles.Preview}>
              <Preview
                previewClass={settings.className}
                template={settings.template}
                styles={this.state.styleValues} />
            </div>
            {
              this.props.isAuth
                ?
                <div className={styles.ButtonContainer}>
                  {
                    this.props.editingItemId &&
                    <a href=""
                      className={styles.SaveButton}
                      onClick={this.saveEdited}>
                      Save
                      </a>
                  }
                  <a href=""
                    className={styles.SaveButton}
                    onClick={this.toggleSaveModal}>
                    Save {this.props.editingItemId && 'As'}
                  </a>
                </div>
                :
                <Link to='/login'
                  className={styles.SaveButton}>
                  Save
              </Link>
            }
          </div>
          <div className={styles.TabsContainer}>
            <Tabs tabs={['Generator', 'Code']}>
              <Generator
                controls={settings.controls}
                controlValues={this.state.controlValues}
                handleChange={this.handleChange}
                onClear={this.clearControlValues} />
              <Code
                previewClass={settings.className}
                template={settings.template}
                styles={this.state.styleValues}
                onCopy={this.openCopyHint} />
            </Tabs>
          </div>
        </div >
        <SaveModal
          isOpen={this.state.isModalOpen}
          onClose={this.toggleSaveModal}
          onSubmit={this.onSave}
        />
      </Auxiliary>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    isAuth: !!state.auth.userId,
    userId: state.auth.userId,
    token: state.auth.token,
    editingItemId: state.editors.editingItemId,
    editingItemTitle: state.editors.editingItemTitle,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    save: (title, editorName, controlValues, userId, token) => {
      dispatch(actions.save(title, editorName, controlValues, userId, token));
    },
    saveEditedItem: (id, title, editorName, controlValues, userId, token) => {
      dispatch(actions.saveEditedItem(id, title, editorName, controlValues, userId, token));
    },
    openHint: (text) => {
      dispatch(actions.openHint(text));
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Editor);