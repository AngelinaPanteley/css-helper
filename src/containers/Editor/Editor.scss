@import '../../variables.scss';

.Editor {
  @include flex-wrapper(row,start, stretch);

  min-height: calc(100vh - #{$header-height});
}

.PreviewContainer {
  @include flex-wrapper(column);

  width: 70%;
  padding: 10px;
}

.Preview {
  width: 80%;
}

.TabsContainer {
  width: 30%;
}

.ToggleLink {
  align-self: flex-start;
  margin: 15px 0;
}