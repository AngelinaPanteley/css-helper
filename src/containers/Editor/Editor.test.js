"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.1';
import { MemoryRouter, Link } from "react-router-dom";
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import Editor from './Editor.js';
import reducer from '../../store/reducers/index';

configure({ adapter: new Adapter() });
const composeEnhancers = process.env.NODE_ENV === 'development' ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : null || compose;

const store = createStore(reducer, composeEnhancers(
  applyMiddleware(thunk)
));

describe('>>>Editor', () => {

  //SNAPSHOTS
  const props = {
    settings: {
      controls: {
        control: {
          type: 'number'
        }
      }
    }
  }

  const editorComponent = <Provider store={store}>
    <MemoryRouter>
      <Editor {...props} />
    </MemoryRouter>
  </Provider>;

  const component = renderer.create(editorComponent);

  it('renders itself correctly', () => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
  });
});
