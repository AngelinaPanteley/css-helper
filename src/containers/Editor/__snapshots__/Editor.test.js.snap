// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>Editor renders itself correctly 1`] = `
Array [
  <div>
    <div>
      <div
        className=" "
      />
      <a
        href=""
        onClick={[Function]}
      >
        Hide Examples
      </a>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
        />
      </div>
      <a
        href="/login"
        onClick={[Function]}
      >
        Save
      </a>
    </div>
    <div>
      <div
        className="Tabs"
        data-tabs={true}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <ul
          className="react-tabs__tab-list"
          role="tablist"
        >
          <li
            aria-controls="react-tabs-1"
            aria-disabled="false"
            aria-selected="true"
            className="react-tabs__tab react-tabs__tab--selected"
            id="react-tabs-0"
            role="tab"
            style={
              Object {
                "width": "calc(50% + 18px)",
              }
            }
            tabIndex="0"
          >
            Generator
          </li>
          <li
            aria-controls="react-tabs-3"
            aria-disabled="false"
            aria-selected="false"
            className="react-tabs__tab"
            id="react-tabs-2"
            role="tab"
            style={
              Object {
                "width": "calc(50% + 18px)",
              }
            }
            tabIndex={null}
          >
            Code
          </li>
        </ul>
        <div
          aria-labelledby="react-tabs-0"
          className="react-tabs__tab-panel react-tabs__tab-panel--selected"
          id="react-tabs-1"
          role="tabpanel"
        >
          <div>
            <button
              onClick={[Function]}
            >
              Clear All
            </button>
            <div>
              <label>
                <span />
                <span />
              </label>
              <div
                className="rc-slider"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                onTouchStart={[Function]}
              >
                <div
                  className="rc-slider-rail"
                  style={Object {}}
                />
                <div
                  className="rc-slider-track"
                  style={
                    Object {
                      "left": "0%",
                      "width": "0%",
                    }
                  }
                />
                <div
                  className="rc-slider-step"
                />
                <div
                  aria-disabled={false}
                  aria-valuemax={100}
                  aria-valuemin={0}
                  aria-valuenow={0}
                  className="rc-slider-handle"
                  onBlur={[Function]}
                  onKeyDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="slider"
                  style={
                    Object {
                      "left": "0%",
                    }
                  }
                  tabIndex={0}
                />
                <div
                  className="rc-slider-mark"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="react-tabs-2"
          className="react-tabs__tab-panel"
          id="react-tabs-3"
          role="tabpanel"
        />
      </div>
    </div>
  </div>,
  <div
    className=" "
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="title"
      >
        Enter title:
      </label>
      <input
        id="title"
        name="title"
        onChange={[Function]}
        type="text"
      />
      <div>
        <button
          onClick={[Function]}
        >
          Cancel
        </button>
        <input
          disabled={true}
          type="submit"
          value="Save"
        />
      </div>
    </form>
  </div>,
]
`;
